src/main/resources/com/nec/cyclone/cpp/cyclone/nullable_scalar_vector.cc

```diff
@@ -373,6 +387,131 @@ const std::vector<size_t> NullableScalarVec<T>::eval_in(const std::vector<T> &el
+  }
+  T* sorted_data = static_cast<T *>(malloc(sizeof(T) * largest_group_size));
+
+  out_group_pos_size = 0;
+  out_group_pos[out_group_pos_size++] = group_pos[0];
+#pragma _NEC vector
+  for(auto g = 1; g < group_pos_size; g++){
+    auto start = group_pos[g - 1];
+    auto end = group_pos[g];
+
+    auto total_el_count = end - start;
+    if(total_el_count == 1){
+      // Shortcut for single element groups
+      if(iter_order_arr == nullptr){
+        idx_arr[start] = start;
+      }else{
+        idx_arr[start] = iter_order_arr[start];
+      }
+      out_group_pos[out_group_pos_size++] = end;
+    }else{
+      size_t cur_invalid_count = 0;
+      size_t cur_valid_count = 0;
+
+      if(iter_order_arr == nullptr){
+#pragma _NEC vector
+#pragma _NEC ivdep
+        for(auto i = start; i < end; i++){
+          if(get_validity(i)){
+            idx_arr[start + cur_valid_count++] = i;
+          }else{
+            idx_arr[end - (++cur_invalid_count)] = i;
+          }
+        }
+      }else{
+#pragma _NEC vector
+#pragma _NEC ivdep
+        for(auto i = start; i < end; i++){
+          auto j = iter_order_arr[i];
+          if(get_validity(j)){
+            idx_arr[start + cur_valid_count++] = j;
+          }else{
+            idx_arr[end - (++cur_invalid_count)] = j;
+          }
+        }
+      }
+
+      { // Setup valid inputs
+#pragma _NEC vector
+        for (auto i = 0; i < cur_valid_count; i++) {
+            sorted_data[i] = data[idx_arr[start + i]];
+        }
+      }
+
+      // Sort data for grouping
+      frovedis::radix_sort(sorted_data, &idx_arr[start], cur_valid_count);
+      std::vector<size_t> group_pos_idxs = frovedis::set_separate(sorted_data, cur_valid_count);
+
+      auto new_group_count = group_pos_idxs.size();
+      auto new_group_arr = group_pos_idxs.data();
+      size_t out_group_idx = out_group_pos_size;
+#pragma _NEC vector
+      for(auto i = 1; i < new_group_count; i++){
+        // We are skipping the first entry here, because it will already be
+        // included in the result, either as the very first value, or because
+        // it was specified as the last value from a previous iteration.
+        auto offset_idx = new_group_arr[i] + start;
+        out_group_pos[out_group_idx++] = offset_idx;
+      }
+      out_group_pos_size = out_group_idx;
+      // The last group index will be based on the last valid group
+      // to account for the invalid group, if it exists, we need to
+      // add the last possible index of this subset, too
+      if(cur_invalid_count > 0){
+        out_group_pos[out_group_pos_size++] = end;
+      }
+    }
+  }
+
+  free(sorted_data);
+}
+
+template <typename T>
+const std::vector<std::vector<size_t>> NullableScalarVec<T>::group_indexes() const {
+  // Short-circuit for simple cases
+  if(count == 0) return {};
+  if(count == 1) return {{0}};
+
+
+  size_t group_pos[2] = {0, static_cast<size_t>(count)};
```

Comment:
I think using `std::vector` for `idx_arr` and `max_group_pos` and  having `group_indexes_on_subset` take in `std::vector` references.  Inside `group_indexes_on_subset` keeps the code more readable and reduces the chances of a missed `free()`.  Hopefully that has no adverse performance effects, since we would fall back to `vector.data()` for vectorization inside `group_indexes_on_subset`...

Reply:
It started out with that interface, but by going entirely std::vector-less I shaved another 2s down from the `many groups` benchmark. In the real world, it also reduces the EDA demo aggregation time by about 2s. 

group_indexes_on_subset is an advanced api that assumes external memory management. That gives us the ability to group on as many keys as we want with constant memory usage.

I do agree with you that it is another foot gun, but as it takes pointers to any external memory it is using, I guess that it should be at least somewhat obvious that you need to free the memory that you've allocated.

Reply:
Gotcha.  It's a bummer that `nc++` isn't able to optimize as well for the `std::vector` case. I'm guessing that keeping `group_indexes_on_subset` interface as is and passing in `vector.data()` as the arg also interferes with optimization.